pkg/chain/chain.go

```diff
@@ -69,95 +68,96 @@ type BondedECDSAKeepFactory interface {
 	// GetKeepCount returns number of keeps.
 	GetKeepCount() (*big.Int, error)
 
-	// GetKeepAtIndex returns the address of the keep at the given index.
-	GetKeepAtIndex(keepIndex *big.Int) (common.Address, error)
+	// GetKeepAtIndex returns a handle to the keep at the given index.
+	GetKeepAtIndex(keepIndex *big.Int) (BondedECDSAKeepHandle, error)
+	// GetKeepWithID returns a handle to the keep with the given ID.
+	// FIXME currently this ID is still a common.Address 😬
+	GetKeepWithID(keepID common.Address) (BondedECDSAKeepHandle, error)
 }
 
-// BondedECDSAKeep is an interface that provides ability to interact with
-// BondedECDSAKeep ethereum contracts.
-type BondedECDSAKeep interface {
+// BondedECDSAKeepHandle is an interface that provides ability to interact with
+// a single bonded ECDSA keep's on-chain component. A bonded ECDSA keep is a
+// threshold signing group that has a corresponding bond amount securing its
+// honest cooperation in the threshold signature application that the keep
+// corresponds to.
+type BondedECDSAKeepHandle interface {
+	// GetKeepWithID returns the underlying keep's ID.
+	// FIXME currently this ID is still a common.Address 😬
+	ID() common.Address
+
 	// OnSignatureRequested installs a callback that is invoked when an on-chain
 	// notification of a new signing request for a given keep is seen.
 	OnSignatureRequested(
-		keepAddress common.Address,
 		handler func(event *SignatureRequestedEvent),
 	) (subscription.EventSubscription, error)
 
 	// OnConflictingPublicKeySubmitted installs a callback that is invoked upon
 	// notification of mismatched public keys that were submitted by keep members.
 	OnConflictingPublicKeySubmitted(
-		keepAddress common.Address,
 		handler func(event *ConflictingPublicKeySubmittedEvent),
 	) (subscription.EventSubscription, error)
 
 	// OnPublicKeyPublished installs a callback that is invoked upon
 	// notification of a published public key, which means that all members have
 	// submitted the same key.
 	OnPublicKeyPublished(
-		keepAddress common.Address,
 		handler func(event *PublicKeyPublishedEvent),
 	) (subscription.EventSubscription, error)
 
 	// SubmitKeepPublicKey submits a 64-byte serialized public key to a keep
 	// contract deployed under a given address.
-	SubmitKeepPublicKey(keepAddress common.Address, publicKey [64]byte) error // TODO: Add promise *async.KeepPublicKeySubmissionPromise
+	SubmitKeepPublicKey(publicKey [64]byte) error // TODO: Add promise *async.KeepPublicKeySubmissionPromise
```

Comment:
I believe that's how it currently works, and I'm comfortable with that. I can drop the TODO in the next PR---I think it's been there a while hehe.

Reply:
I think it's great though, I meant to ask about a couple of these TODOs. Have to clean them up when we notice them otherwise they stick around forever haha.

Reply:
Dropped.